From 902f3c0fdf8c95f93ec617aa750eaefe3c3e53fd Mon Sep 17 00:00:00 2001
From: Zdenek Dohnal <zdohnal@redhat.com>
Date: Mon, 3 Dec 2018 15:09:50 +0100
Subject: [PATCH 06/11] udev/udev-configure-printer.c: device_uri strings leaks

---
 udev/udev-configure-printer.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/udev/udev-configure-printer.c b/udev/udev-configure-printer.c
index ccb9b78a..fc0676f0 100644
--- a/udev/udev-configure-printer.c
+++ b/udev/udev-configure-printer.c
@@ -1512,6 +1512,8 @@ for_each_matching_queue (struct device_uris *device_uris,
       firstqueue = 0;
 
     skip:
+      free(device_uri_n);
+      free(this_device_uri_n);
       if (!attr)
 	break;
     }
-- 
2.20.1

