From 04ddf2fba0d9dd84c29430d688ceca1ae500885c Mon Sep 17 00:00:00 2001
From: Zdenek Dohnal <zdohnal@redhat.com>
Date: Thu, 3 Jan 2019 10:11:27 +0100
Subject: [PATCH 09/11] Temporary fix for error pop-up window for IPP2.0+
 attributes

---
 NEWS                 | 1 +
 printerproperties.py | 7 +------
 2 files changed, 2 insertions(+), 6 deletions(-)

diff --git a/printerproperties.py b/printerproperties.py
index ff85664b..ca43c25e 100755
--- a/printerproperties.py
+++ b/printerproperties.py
@@ -1430,12 +1430,7 @@ class PrinterPropertiesDialog(GtkGUI):
                 except:
                     nonfatalException()
                     option_editable = False
-                    show_error_dialog (_("Error"),
-                                       _("Option '%s' has value '%s' and "
-                                         "cannot be edited.") %
-                                       (option.name,
-                                        value),
-                                       self.parent)
+                    debugprint ("Option '%s' has value '%s' and cannot be edited." % (option.name, value))
             option.widget.set_sensitive (option_editable)
             if not editable:
                 option.button.set_sensitive (False)
-- 
2.20.1

