diff --git a/system-config-printer.py b/system-config-printer.py
index d4ee88e..ccedf1c 100755
--- a/system-config-printer.py
+++ b/system-config-printer.py
@@ -43,6 +43,7 @@ def show_help():
            "Options:\n\n"
            "  --debug                 Enable debugging output.\n"
            "  --show-jobs <printer>   Show the print queue for <printer>\n"
+           "  --embedded              Enable to start in Embedded mode.\n "
            "  --help                  Show this message.\n")
 
 if len(sys.argv)>1 and sys.argv[1] == '--help':
@@ -90,6 +91,9 @@ pkgdata = config.pkgdatadir
 iconpath = os.path.join (pkgdata, 'icons/')
 sys.path.append (pkgdata)
 
+PlugWindow = None
+PlugWindowId = None
+
 def CUPS_server_hostname ():
     host = cups.getServer ()
     if host[0] == '/':
@@ -192,6 +196,16 @@ class GUI(GtkGUI):
 
                         domain=config.PACKAGE)
 
+        if PlugWindowId:
+            self.PrintersWindow.hide()
+            # the "vbox4" widget
+            vbox = self.PrintersWindow.get_children()[0]
+            PlugWindow = gtk.Plug(PlugWindowId)
+            gtk.Container.remove(self.PrintersWindow, vbox)
+            PlugWindow.add(vbox)
+            self.PrintersWindow.set_transient_for(PlugWindow)
+            PlugWindow.show_all()
+            self.PrintersWindow = PlugWindow
 
         # Since some dialogs are reused we can't let the delete-event's
         # default handler destroy them
@@ -2101,7 +2115,8 @@ if __name__ == "__main__":
     import getopt
     try:
         opts, args = getopt.gnu_getopt (sys.argv[1:], '',
-                                        ['debug', 'show-jobs='])
+                                        ['embedded=', 
+                                            'debug', 'show-jobs='])
     except getopt.GetoptError:
         show_help ()
         sys.exit (1)
@@ -2115,4 +2130,7 @@ if __name__ == "__main__":
         elif opt == '--show-jobs':
             show_jobs = optarg
 
+        if opt == "--embedded":
+            PlugWindowId = long(optarg)
+
     main(show_jobs)
