--- system-config-printer-1.1.12/system-config-printer.py
+++ system-config-printer-1.1.12/system-config-printer.py
@@ -135,6 +135,9 @@ sys.path.append (pkgdata)
 
 busy_cursor = gtk.gdk.Cursor(gtk.gdk.WATCH)
 
+PlugWindow = None
+PlugWindowId = None
+
 TEXT_start_firewall_tool = _("To do this, select "
                              "System->Administration->Firewall "
                              "from the main menu.")
@@ -357,6 +360,17 @@ class GUI(GtkGUI, monitor.Watcher):
         # the --no-focus-on-map option
         self.PrintersWindow.set_focus_on_map (self.focus_on_map)
 
+        if PlugWindowId:
+            self.PrintersWindow.hide()
+            # the "vbox4" widget
+            vbox = self.PrintersWindow.get_children()[0]
+            PlugWindow = gtk.Plug(PlugWindowId)
+            gtk.Container.remove(self.PrintersWindow, vbox)
+            PlugWindow.add(vbox)
+            self.PrintersWindow.set_transient_for(PlugWindow)
+            PlugWindow.show_all()
+            self.PrintersWindow = PlugWindow
+
         # Since some dialogs are reused we can't let the delete-event's
         # default handler destroy them
         for dialog in [self.PrinterPropertiesDialog,
@@ -6809,6 +6822,7 @@ if __name__ == "__main__":
                                          'devid=',
                                          'print-test-page=',
                                          'no-focus-on-map',
+                                         'embedded=',
                                          'debug'])
     except getopt.GetoptError:
         show_help ()
@@ -6841,6 +6855,8 @@ if __name__ == "__main__":
 
         elif opt == '--debug':
             set_debugging (True)
+        if opt == "--embedded":
+            PlugWindowId = long(optarg)
 
     main(setup_printer, configure_printer, change_ppd, devid, print_test_page,
          focus_on_map)
