--- system-config-printer-1.3.11/system-config-printer.py.orig	2012-08-01 11:48:26.000000000 -0400
+++ system-config-printer-1.3.11/system-config-printer.py	2012-08-19 12:00:21.112797099 -0400
@@ -47,7 +47,8 @@ def show_help():
            "a CUPS server configuration program.\n\n"
            "Options:\n\n"
            "  --debug                 Enable debugging output.\n"
-           "  --show-jobs <printer>   Show the print queue for <printer>\n")
+           "  --show-jobs <printer>   Show the print queue for <printer>\n"
+           "  --embedded              Enable to start in Embedded mode.\n " )
 
 if len(sys.argv)>1 and sys.argv[1] == '--help':
     show_help ()
@@ -97,6 +98,9 @@ pkgdata = config.pkgdatadir
 iconpath = os.path.join (pkgdata, 'icons/')
 sys.path.append (pkgdata)
 
+PlugWindow = None
+PlugWindowId = None
+
 def CUPS_server_hostname ():
     host = cups.getServer ()
     if host[0] == '/':
@@ -199,7 +203,17 @@ class GUI(GtkGUI):
 
                         domain=config.PACKAGE)
 
-
+        if PlugWindowId:
+            self.PrintersWindow.hide()
+            # the "vbox4" widget
+            vbox = self.PrintersWindow.get_children()[0]
+            PlugWindow = gtk.Plug(PlugWindowId)
+            gtk.Container.remove(self.PrintersWindow, vbox)
+            PlugWindow.add(vbox)
+            self.PrintersWindow.set_transient_for(PlugWindow)
+            PlugWindow.show_all()
+            self.PrintersWindow = PlugWindow
+		
         # Since some dialogs are reused we can't let the delete-event's
         # default handler destroy them
         self.ConnectingDialog.connect ("delete-event",
@@ -1993,7 +2007,8 @@ if __name__ == "__main__":
     import getopt
     try:
         opts, args = getopt.gnu_getopt (sys.argv[1:], '',
-                                        ['debug', 'show-jobs='])
+                                        ['embedded=',
+                                         'debug', 'show-jobs='])
     except getopt.GetoptError:
         show_help ()
         sys.exit (1)
@@ -2006,5 +2021,7 @@ if __name__ == "__main__":
             cupshelpers.ppds.set_debugprint_fn (debugprint)
         elif opt == '--show-jobs':
             show_jobs = optarg
+        if opt == "--embedded":
+            PlugWindowId = long(optarg)
 
     main(show_jobs)
